<a href="https://colab.research.google.com/github/lkoenigs-2025/Module-6/blob/main/SP_26_MCON_232_Classes_and_Objects_Quiz_LK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FC Classes and Objects Quiz  
**Module: Classes and Objects in Python**  

## Instructions
- This quiz is **untimed**.
- For a **fast student**, this quiz is designed to take about **60 minutes**.
- If you need to stop in order to **study**, that is allowed. However, please wait **3 hours** before resuming the quiz.
- If you need a break for any reason **other than study**, you may resume the quiz **immediately** after your break.
- **Use of AI is forbidden.**
- Please complete your own work and show what you know.

## Short reminders before you begin
A **class** is a blueprint.  
An **object** is one instance created from that blueprint.  
The `__init__` method builds the starting state of each object.  
An **instance attribute** belongs to one object, usually written with `self.attribute_name`.  
A **class attribute** belongs to the class itself and is shared unless shadowed by an instance.  
A **method** is a function inside a class.  
When an object is passed into a method, the method receives a reference to that object.  
If the method **mutates the object's attributes**, the caller sees those changes after the method ends.  
If the method **reassigns the parameter variable** to a different object, that reassignment only changes the local variable inside the method.

## Academic Integrity
By submitting this notebook, you are affirming that this work is your own and that you did not use AI or outside help that is not permitted.

## Requirements
- for any print statement `__str__` is required
- all parameters must have a data type associated with it, as well as return type from method.
- please comment ALL logic (otherwise comments are not necessary)

## Quick Reference
Keep these ideas in mind while working:

- Define a class with `class ClassName:`
- Build objects with `obj = ClassName(...)`
- Instance methods usually begin with `self`
- Access instance data with `self.attribute`
- Access class data with `ClassName.attribute`
- You can also read a class attribute through an object, but be careful: assigning through the object can create an instance attribute that shadows the class attribute
- Passing an object into another object's method allows one object to use the other object's data
- Mutation persists after the method call
- Reassignment of the parameter name does not replace the original caller's object

In [27]:

# Run this cell first.
# You may use it for quick checks throughout the quiz.

def assert_equal(actual, expected, label=""):
    if actual == expected:
        print(f"PASS: {label}")
    else:
        print(f"FAIL: {label}\n  expected: {expected}\n  got:      {actual}")

# Part A — Concepts and Vocabulary

Answer in full sentences. Keep each answer to about **2–4 sentences**.

### Question A1
What is the difference between a **class** and an **object**?

**Your answer:**
A class is a blueprint. An object is one instance created from the blueprint.

### Question A2
What is the role of the `__init__` method?

**Your answer:**
The init method intializes/starts an object from the class. It uses self.attribute.

### Question A3
What is the difference between an **instance attribute** and a **class attribute**?

**Your answer:**
An instance object is assigned to a specific object. A class object is assigned to the whole class- evey object made from that blueprint.

### Question A4
When an object is passed to another object's method, what is actually being passed?

**Your answer:**
It references the class from the other object inside the objects method.

### Question A5
What is the difference between **mutating an object** and **reassigning a parameter variable** inside a method?

**Your answer:**
When you mutate an object the caller sees it when the method ends. When reassigning a parameter variable it changes the local variable inside the method but the change is not seen by the object.

# Part B — Reading Code

For each question:
1. Read the code carefully.
2. Predict the output or final object state.
3. Then run the code cell if you want to check your thinking.

### Question B1 — Basic object state

Before running the code, answer:
- What are the values of `name` and `score` in `player1` after the object is created?

In [3]:

class Player:
    def __init__(self, name, score):
        self.name = name
        self.score = score

player1 = Player("Avi", 10)

print(player1.name)
print(player1.score)

Avi
10


**Your answer:**
name = "Avi,
score = 10,
player_1 = one instance of the class Player with name as "Avi" and score as 10

### Question B2 — Class variable vs instance variable

Before running the code, answer:
- What is printed?
- Why?

In [4]:

class Counter:
    total_created = 0

    def __init__(self):
        Counter.total_created += 1

a = Counter()
b = Counter()
c = Counter()

print(Counter.total_created)

3


**Your answer:**
The answer will be 3 beacause Counter.total_created refers to the class attibute and each time a counter object is created it adds one to the total created attribute. Total will be 3.

### Question B3 — Passing one object into another object's method

Before running the code, answer:
- What is printed?
- Which object's data is being used?

In [5]:

class Student:
    def __init__(self, name, grade):
        self.name = name
        self.grade = grade

class Reporter:
    def show_student(self, student_obj):
        print(student_obj.name, student_obj.grade)

s1 = Student("Leah", 95)
r = Reporter()
r.show_student(s1)

Leah 95


**Your answer:**
Printed: Leah 95

The reporter class is using the student class's data to print using the method show student

### Question B4 — Mutation inside a method

Before running the code, answer:
- What is printed at the end?
- Why does that change remain after the method call?

In [6]:

class BankAccount:
    def __init__(self, owner, balance):
        self.owner = owner
        self.balance = balance

class BankTeller:
    def deposit_bonus(self, account, amount):
        account.balance += amount

acct = BankAccount("Mina", 100)
teller = BankTeller()
teller.deposit_bonus(acct, 25)

print(acct.balance)

125


**Your answer:**
Printed: 125

The class object is created using the BankAccount class and then the teller class adds onto the balance by calling the object from the other class(acct). This method changes are seen because it mutates the objects attribute.

### Question B5 — Reassigning the parameter inside a method

Before running the code, answer:
- What is printed?
- Why does the original object stay the same?

In [7]:

class Box:
    def __init__(self, value):
        self.value = value

class Changer:
    def replace_box(self, box_obj):
        box_obj = Box(999)
        box_obj.value += 1

b = Box(10)
c = Changer()
c.replace_box(b)

print(b.value)

10


**Your answer:**
Printed: 10

The object is reassigned using a different class. This change not to the actual object because it only changes the local variable inside the method.

# Part C — Write the Code

Write each class or method in the code cells below.

### Question C1 — Create a simple class

Create a class named `Book`.

#### Requirements
- The constructor should accept:
  - `title`
  - `author`
  - `pages`
- Store each value in an instance attribute.
- Then create an object named `b1` with your own sample data.
- Print the title and author on one line.
- use `__str__`

**Reminder:** Each object should store its own instance data.

In [28]:

# Write your answer here
class Book:
  def __init__(self, title: str, author: str, pages: int) -> None:
    #instance attributes
    self.title = title
    self.author = author
    self.pages = pages

  #print title and author using str function
  def __str__(self) -> str:
    return f"Title of the book: {self.title} \nAuthor: {self.author}"

#create object
b1 = Book("Down the Hill", "Jane Mendel", 240)
print(b1)


Title of the book: Down the Hill 
Author: Jane Mendel


### Question C2 — Add an instance method

Create a class named `Dog`.

#### Requirements
- The constructor should accept:
  - `name`
  - `age`
- Store them as instance attributes.
- Add a method named `describe()` that returns a string like:
  - `"Fido is 4 years old."`
- Create one object and test your method.

**Reminder:** Instance methods use `self` to access that object's data.

In [29]:

# Write your answer here
class Dog:
  def __init__(self, name: str, age: int) -> None:
    #instance attributes
    self.name = name
    self.age = age

  #add method that retuns string
  def describe(self) -> str:
    return f"{self.name} is {self.age} years old."

#create object and test method
d1 = Dog("Fido", 4)
print(d1.describe())

Fido is 4 years old.


### Question C3 — Class variable practice

Create a class named `Student`.

#### Requirements
- Add a class attribute named `school_name` and set it to `"Touro"`.
- The constructor should accept:
  - `name`
  - `major`
- Store them as instance attributes.
- Create two student objects.
- Print each student's `name`.
- Print the shared class variable.

**Reminder:** The class variable belongs to the class and is shared.

In [30]:

# Write your answer here
class Student:
  #class attribute
  school_name = "Touro"
  def __init__(self, name: str, major: str) -> None:
    #instance attributes
    self.name = name
    self.major = major

  #print name and shaed class variable using str function
  def __str__(self) -> str:
    return f"Student of {Student.school_name}: {self.name}"

#two objects
s1 = Student("Leah", "Computer Science")
s2 = Student("Ruchama", "Computer Science")

#print each name and shared variable
print(s1)
print(s2)


Student of Touro: Leah
Student of Touro: Ruchama


### Question C4 — One object uses another object

Create two classes: `Player` and `Coach`.

#### `Player`
- constructor should accept:
  - `name`
  - `points`

#### `Coach`
- add a method `praise_player(player_obj)` that returns a string like:
  - `"Great job, Maya! You scored 18 points."`

Create a player object and a coach object, then call the method.

**Reminder:** This question practices passing an object into another object's method and using its data.

In [31]:

# Write your answer here
class Player:
  def __init__(self, name: str, points: int) -> None:
    self.name = name
    self.points = points

class Coach:
  #method that returns a string
  def praise_player(self, player_obj: Player) -> str:
    self.player_obj = player_obj
    return f"Great job, {player_obj.name}! You scored {player_obj.points} points."

#create objects from class
p1 = Player("Leah", 35)
c1 = Coach()

print(c1.praise_player(p1))


Great job, Leah! You scored 35 points.


### Question C5 — Mutation vs reassignment

Create two classes: `Pet` and `Groomer`.

#### `Pet`
- constructor should accept:
  - `name`
  - `clean`

#### `Groomer`
Add two methods:

1. `wash_pet(pet_obj)`  
   - set `pet_obj.clean` to `True`

2. `replace_pet(pet_obj)`  
   - reassign `pet_obj` to a brand new `Pet("Temporary", True)`

#### After writing the classes
- Create one pet object that starts dirty.
- Call `wash_pet(...)`
- Print the original pet's `name` and `clean` value
- Then call `replace_pet(...)`
- Print the original pet's `name` and `clean` value again

#### Finally, answer in a comment:
- Which method changed the original object?
- Which method did not?
- Why?

**Reminder:** Mutation changes the object itself. Reassignment only changes the local parameter name.

In [32]:
# Write your answer here
class Pet:
  def __init__(self, name: str, clean: bool) -> None:
    self.name = name
    self.clean = clean

class Groomer:
  #sets clean to true
  def wash_pet(self, pet_obj: Pet) -> None:
    self.pet_object = pet_obj
    pet_obj.clean = True

  #reassigns variable to new object
  def replace_pet(self, pet_obj: Pet) -> None:
    pet_obj = Pet("Temporary", True)

#object that starts dirty
p1 = Pet("Fido" , False)
g1 = Groomer()

#wash pet method called and values printed
g1.wash_pet(p1)
print(f"{p1.name} is {p1.clean}")

#replace pet method called and values printed
g1.replace_pet(p1)
print(f"{p1.name} is {p1.clean}")

#Wash pet method changed the object because it is mutating inside the class.
#replace pet is reassigning the object to a local variable parameter so doesnt change when the method is called.


Fido is True
Fido is True


# Part D — Wrap-Up Reflection

Answer in full sentences.

### Question D1
What is one thing you understand better now about classes and objects?

Mutating and reassigning objects.

### Question D2
What is still a little confusing or something you would like to practice more?

I get a little confused with using the self... but otherwise I'm BH getting the hang of it.

## Not graded :  
- How do you feel about this topic ?

I'm really getting the idea!!

- How long did this quiz take you (excluding breaks)

40 min

# Thank you !